### ISSUE TITLE
parser: Shift-reduce conflict between TypeName and Locals

### ISSUE THREAD

**Issue by mewmew** [mewmew/uc#27]
From `LR1_conflicts.txt` generated by Gocc:

```
11 LR-1 conflicts: 
	S28
		symbol: ident
			Reduce(19:Locals : empty	<<  >>)
			Shift(6)
	S35
		symbol: ident
			Reduce(19:Locals : empty	<<  >>)
			Shift(6)
```

Relevant production rules from [the BNF grammar](https://github.com/mewmew/uc/blob/9cae1a7f40de433a4c95194097f55b372e91c669/gocc/uc.bnf):

```
TypeName
	: ident // TypeName : "char" | "int" | "void" ;
;

Locals
	: empty
	| VarDecl ";" Locals
;
```

Example which fails to parse using `uparse` on input [testdata/quiet/parser/p01.c](https://github.com/mewmew/uc/blob/dev/testdata/quiet/parser/p01.c):

```
u@x1 ~/D/g/s/g/m/uc> uparse testdata/quiet/parser/p01.c 
Parsing "testdata/quiet/parser/p01.c"

…

pos: 28
typ: 4
lit: int

S35 ident(4,int) shift:6
pos: 32
typ: 4
lit: y

S6 ident(4,y) reduce:10(TypeName : ident	<<  >>)
S33 ident(4,y) shift:36
pos: 33
typ: 3
lit: ;

S36 ;(3,;) reduce:8(ScalarDecl : TypeName ident	<<  >>)
S7 ;(3,;) reduce:6(VarDecl : ScalarDecl	<<  >>)
S32 ;(3,;) shift:35
pos: 37
typ: 4
lit: x

S35 ident(4,x) shift:6
pos: 39
typ: 29
lit: =

2016/03/31 03:11:06 main.parseFile (uparse.go:69): error: Error in S6: =(29,=), Pos(offset=39, line=7, column=8), expected one of: ident 
```

A potential solution may be to define TypeName as:

```c
TypeName : "char" | "int" | "void" ;
```

This solves the shift-reduce conflict. The decision may have to be revisited when adding support for user-defined types.

**Comment by mewmew** [comment]
Consider reverting commit 721f18d8ad1ab3133fe8b1712ad4f6f2c7c2fb3e when introducing support for user-defined types.

**Comment by mewmew** [comment]
If we could find a better solution for this issue, that would be preferable, as we already know it may have to be revisited once user-defined types are introduced.

An alternative solution could be to remove the `TypeName` production rule entirely, and use `ident` instead of `TypeName` in the other production rules. This would remove the `TypeName` reduce, and keep `ident`s on the stack, which should resolve the shift-reduce conflict. This is very much a theory for now, as it has not yet been tested, and I still feel a lack of intuition for how the LR-1 parser will handle this situation once the alternative solution has been applied to the grammar. For now, merely pointing it out so that we may evaluate it later on.

Re-opening this issue until we take a proper decision on which route to venture down.

**Comment by mewmew** [comment]
Another solution may be to remove the `Locals` production rule and add a `DeclStmt` production rule for declaration statements to the `Stmt` production rule. This would allow declaration statements to occur anywhere within blocks, as is valid for later versions of C (with some constraints related to `goto`, i.e. cannot jump over declaration).

I propose that we remove the `Locals` production rule and allow declarations to occur anywhere a statement may occur. To enforce uC behaviour we could add a check during the semantic analysis phase which verified that `DeclStmt`s only occur before any other statements within the body of functions.

This would effectively solve the The Hack Issue, as we may revert the `TypeName` production rule to `ident` rather than hardcoded `"char" | "int" | "void"`, and as such, this would allow for future extensions with regards to user-defined data types.

Furthermore, we may want to take this approach in the future either way, as our intention is to support a larger and more modern subset of C in the future. Being able to declare variables closer to their initial user is a major win for readability and maintainability of code.

Alex, I'd be interested to hear your thoughts on this. Should we go ahead and remove the `Locals` production rule?

**Comment by mewmew** [comment]
> Alex, I'd be interested to hear your thoughts on this. Should we go ahead and remove the Locals production rule?

I'll try to implement this and see if it results in cleaner code. We can revert later once we have the two approaches to compare against one another.

**Comment by mewmew** [comment]
With commit f7c2633 we are back to using ident for the `TypeName` production rule, thus handling the Lexer Hack in the same way as Clang. Closing this issue for now. We can easily re-open it later, if after discussion we would prefer to solve this issue in some other way.

**Comment by sangisos** [comment]
No that sounds good to me, that was the way we said we would like to do it
anyway. Nice work.

Regards
Alexander Andersson
On 10 Apr 2016 02:52, "Robin Eklind" <notifications@github.com> wrote:

Closed #27 <https://github.com/mewmew/uc/issues/27>.

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly or view it on GitHub
<https://github.com/mewmew/uc/issues/27#event-621308056>
